File: src/main/java/com/lessmarkup/framework/helpers/TypeHelper.java
package com.lessmarkup.framework.helpers;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.*;

public class TypeHelper {
    
    private static final Map<Class<?>, Collection<PropertyDescriptor>> propertyMap = new HashMap<>();

    public synchronized static Iterable<PropertyDescriptor> getProperties(Class<?> type) {
        
        Collection<PropertyDescriptor> ret = propertyMap.get(type);
        if (ret != null) {
            return ret;
        }
        
        ret = new ArrayList<>();
        propertyMap.put(type, ret);

        Map<String, Method> methods = new HashMap<>();

        for (Method method : type.getMethods()) {
            if ((method.getModifiers() & Modifier.STATIC) != 0) {
                continue;
            }
            methods.put(method.getName(), method);
        }

        for (Map.Entry<String, Method> entry : methods.entrySet()) {
            if (!entry.getKey().startsWith("set")) {
                continue;
            }

            Method setter = entry.getValue();

            if (setter.getParameterCount() != 1) {
                continue;
            }

            Class<?> fieldType = setter.getParameterTypes()[0];

            String propertyName = entry.getKey().substring(3);
            String getterName;

            if (boolean.class.equals(fieldType)) {
                getterName = "is" + propertyName;
            } else {
                getterName = "get" + propertyName;
            }

            Method getter = methods.get(getterName);

            if (getter == null) {
                continue;
            }

            ret.add(new PropertyDescriptor(StringHelper.toJsonCase(propertyName), getter, setter, fieldType));
        }

        return ret;
    }
}

File: src/main/scala/com/lessmarkup/framework/helpers/TypeHelper.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.framework.helpers

import java.lang.reflect.Modifier
import scala.collection.concurrent

object TypeHelper {

  private val propertyMap: concurrent.Map[Class[_], Seq[PropertyDescriptor]] = concurrent.TrieMap()

  def getProperties(dataType: Class[_]): Seq[PropertyDescriptor] = {

    val ret = propertyMap.get(dataType)
    if (ret.isDefined) {
      return ret.get
    }

    val methods = dataType.getMethods
      .filter(m => !Modifier.isStatic(m.getModifiers))
      .map(m => (m.getName, m))
      .toMap

    val properties: Seq[PropertyDescriptor] = (for (
      (key, setter) <- methods
      if key.endsWith("_$eq") && setter.getParameterCount == 1;
      fieldType = setter.getParameterTypes()(0);
      propertyName = key.substring(0, key.length-4);
      getter = methods.get(propertyName)
      if getter.isDefined
    ) yield new PropertyDescriptor(StringHelper.toJsonCase(propertyName), getter.get, setter, fieldType)
    ).toSeq

    propertyMap.put(dataType, properties)
    properties
  }
}
